be/src/exec/pipeline/nljoin/spillable_nljoin_build_operator.cpp

```diff
@@ -0,0 +1,114 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/pipeline/nljoin/spillable_nljoin_build_operator.h"
+
+#include "exec/pipeline/nljoin/nljoin_build_operator.h"
+#include "exec/pipeline/query_context.h"
+#include "exec/spill/options.h"
+#include "exec/spill/spiller.hpp"
+#include "gen_cpp/InternalService_types.h"
+
+namespace starrocks::pipeline {
+Status SpillableNLJoinBuildOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(NLJoinBuildOperator::prepare(state));
+    _spill_channel->spiller()->set_metrics(spill::SpillProcessMetrics(_unique_metrics.get()));
+    RETURN_IF_ERROR(_spill_channel->spiller()->prepare(state));
+    _cross_join_context->input_channel(_driver_sequence).set_spiller(_spill_channel->spiller());
+    if (state->spill_mode() == TSpillMode::FORCE) {
+        _strategy = spill::SpillStrategy::SPILL_ALL;
+    }
+    return Status::OK();
+}
+
+void SpillableNLJoinBuildOperator::close(RuntimeState* state) {
+    NLJoinBuildOperator::close(state);
+}
+
+bool SpillableNLJoinBuildOperator::need_input() const {
+    if (_strategy == spill::SpillStrategy::NO_SPILL) {
+        return NLJoinBuildOperator::need_input();
+    }
+    return !_is_finished && !_spill_channel->spiller()->is_full();
+}
+
+bool SpillableNLJoinBuildOperator::is_finished() const {
+    if (!_spill_channel->spiller()->spilled()) {
+        return NLJoinBuildOperator::is_finished();
+    }
+    return _is_finished;
+}
+
+Status SpillableNLJoinBuildOperator::set_finishing(RuntimeState* state) {
+    auto& executor = *_spill_channel->io_executor();
+    auto spiller = _spill_channel->spiller();
+
+    if (!spiller->spilled()) {
+        _spill_channel->set_finishing();
+        RETURN_IF_ERROR(NLJoinBuildOperator::set_finishing(state));
+        return Status::OK();
+    }
+
+    RETURN_IF_ERROR(spiller->flush(state, executor, RESOURCE_TLS_MEMTRACER_GUARD(state)));
+    spiller->set_flush_all_call_back(
+            [&, state]() {
+                RETURN_IF_ERROR(_cross_join_context->finish_one_right_sinker(_driver_sequence, state));
+                _is_finished = true;
+                _spill_channel->set_finishing();
+                return Status::OK();
+            },
+            state, executor, RESOURCE_TLS_MEMTRACER_GUARD(state));
+
+    return Status::OK();
+}
+
+Status SpillableNLJoinBuildOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    if (_strategy == spill::SpillStrategy::NO_SPILL) {
+        RETURN_IF_ERROR(NLJoinBuildOperator::push_chunk(state, chunk));
+    } else {
+        // TODO: process auto spill mode
+        _cross_join_context->input_channel(_driver_sequence)
```

Comment:
is it always spilling here? what if build side only has a few rows?

Reply:
if build side only has a few rows. it won't really flush to disk. it will only cached in memory table.